### PR TITLE
[compiler] ensure CDA result Array[Array[Byte]] can be GCed after dec…

### PR DESCRIPTION
**Pull request by tpoterba**
…oding

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 1]
```diff
@@ -2498,6 +2482,9 @@ class Emit[C](
               cb.assign(stageName, stageName.concat("|").concat(dynamicID.asString.loadString(cb)))
             })
 
+          val encRes = cb.newLocal[Array[Array[Byte]]]("encRes")
+          val len = mb.newLocal[Int]("cda_result_length")
+          val ib = mb.newLocal[InputBuffer]("decode_ib")
```
Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala:2487]: why sometimes class builder sometimes method builder?

Comment (danking) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala:2487]: More direct: How do you pick when to use mb vs cb?

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 2]
```diff
@@ -2508,7 +2495,18 @@ class Emit[C](
...
+          cb.assign(len, encRes.length())
+          val pt = PCanonicalArray(bodySpec.encodedType.decodedSType(bodySpec.encodedVirtualType).asInstanceOf[SBaseStruct].fieldEmitTypes(0).storageType)
+          val resultArray = pt.constructFromElements(cb, region, len, deepCopy = false) { (cb, i) =>
+            cb.assign(ib, bodySpec.buildCodeInputBuffer(Code.newInstance[ByteArrayInputStream, Array[Byte]](encRes(i))))
+            val eltTupled = bodySpec.encodedType.buildDecoder(bodySpec.encodedVirtualType, parentCB)
+              .apply(cb, region, ib)
+              .asBaseStruct
+            eltTupled.loadField(cb, 0)
+          }
+          cb.assign(encRes, Code._null)
```
Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala:2508]: Is it not possible to also free each block after decoding it? As written we'd need twice the RAM to read a given result array, I think.

**Review by danking — DISMISSED** [review]

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala:2487]: These are aliased. Rewrote to use the more idiomatic "memoize" now and avoid explicit variable declarations.

The rule is basically to use the lowest-level builder that has the functionality you want. Module > Class > Method > Code, but codeBuilder doesn't duplicate the full functionality of everything above. So to create a field, use methodBuilder (it exists on MB and ClassB, not codeB).

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala (hunk 2 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala:2508]: nice

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 3]
```diff
@@ -2498,6 +2482,7 @@ class Emit[C](
               cb.assign(stageName, stageName.concat("|").concat(dynamicID.asString.loadString(cb)))
             })
 
+          val encRes = cb.newLocal[Array[Array[Byte]]]("encRes")
           cb.assign(encRes, spark.invoke[BackendContext, HailClassLoader, FS, String, Array[Array[Byte]], Array[Byte], String, Option[TableStageDependency], Array[Array[Byte]]](
```
Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala:2486]: Approved, but is it possible to use `memoize` here as well or is this special in some way?

**Pull request merged**